text/0000-opentelemetry-protocol.md

```diff
@@ -0,0 +1,460 @@
+
+## Prior Art, Alternatives and Future Possibilities
+
+OTLP is an evolution of OpenCensus protocol based on the research and testing of its modifications in production at Omnition. The modifications include the addition of delivery acknowledgements and changes that make the protocol load-balancer friendly.
+
+OTLP uses Protocol Buffers for data encoding. Two other encodings were considered as alternative approaches: FlatBuffers and Capnproto. Both alternatives were rejected. FlatBuffers was rejected because it lacks required functionality in all languages except C++, particularly lack of verification of decoded data and inability to mutate in-memory data. Capnproto was rejected because it is not yet considered production ready, the API is not yet stable and like FlatBuffers it lacks ability mutate in-memory data.
+
+Both FlatBuffers and Capnproto are worth to be re-evaluated for future versions of OpenTelemetry protocol if they overcome currently known limitations.
+
+It is also worth researching transports other than gRPC. Other transports are not included in this RFC due to time limitations.
+
+Experimental implementation of OTLP over WebSockets exists and was researched as an alternate. WebSockets were not chosen as the primary transport for OTLP due to lack or immaturity of certain capabilities (such as [lack of universal support](https://github.com/gorilla/websocket#gorilla-websocket-compared-with-other-packages) for [RFC 7692](https://tools.ietf.org/html/rfc7692) message compression extension). Despite limitations the experimental implementation demonstrated good performance and WebSocket transport will be considered for inclusion in a future OTLP Extensions RFC.
+
+## Open Questions
+
+One of the goals for telemetry protocol is reducing CPU usage and memory pressure in garbage collected languages. These goals were not addressed as part of this RFC and remain open. One of the promising future ways to address this is finding a more CPU and memory efficient encoding mechanism.
+
+Another goal for telemetry protocol is achieving high compression ratios for telemetry data while keeping CPU consumption low. OTLP uses compression provided by gRPC transport. No further improvements to compression were considered as part of this RFC and are a future area of work.
+
+## Appendix A - Protocol Buffer Definitions
+
+This is Protocol Buffers schema for `Export` request and response:
+
+```
+// Hello is the first request from client to server.
+message HelloRequest {
+    // Empty in initial version of OTLP.
+}
+
+// Response to Hello request.
+message HelloResponse {
+    // Empty in initial version of OTLP.
+}
+
+// A request from client to server containing telemetry data to export.
+message ExportRequest {
+  // Unique sequential ID generated by the client.
+  uint64 id = 1;
+
+  // Telemetry data.
+  repeated NodeSpans nodeSpans = 2;
+  repeated NodeMetrics nodeMetrics = 3;
```

Comment:
Can you please clarify why?

The reason I placed different telemetry data types into one ExportRequest is because this makes client and server implementations at the transport level (level of this spec) more uniform. The client and server do not care about telemetry data type, the exporting protocol is the same regardless of telemetry data type.

If each telemetry data had a different message type and a different gRPC service call it would create unnecessary duplication of code and logic in the implementation.

Perhaps you have something else in your mind, please elaborate.

Reply:
Usually metrics have different requirements in terms of retries/load-balancing/reporting periods etc. Having them in the same protocol forces us to apply the same configuration, or at least is less intuitive. Also even in the Receivers we would probably have different configurations for metrics/spans.

Also if any of the OSS backends like Jaeger/Zipkin decide to support this the will support half of the "Service". Having them in different grpc Services makes it more clear that Jaeger supports TraceService (so it supports the entire service).

Reply:
OK, admittedly I spent less time thinking about metrics.

Can you please expand a bit more on this? How is metrics requirements different? I want to understand if different requirements will mean that at the protocol level we need to do something different too (e.g. do we need acks or we don't). You have more experience with this so your thoughts are very useful.